features/manage-position/Withdraw.tsx

```diff
@@ -0,0 +1,342 @@
+
+  const [collateralToWithdraw, setCollateralToWithdraw] = useState<string>("");
+  const [hash, setHash] = useState<string | null>(null);
+  const [success, setSuccess] = useState<boolean | null>(null);
+  const [error, setError] = useState<Error | null>(null);
+
+  const withdrawCollateral = async () => {
+    if (collateralToWithdraw && emp) {
+      setHash(null);
+      setSuccess(null);
+      setError(null);
+      const collateralToWithdrawWei = ethers.utils.parseUnits(
+        collateralToWithdraw
+      );
+      try {
+        if (resultingCRBelowGCR) {
+          const tx = await emp.requestWithdrawal([collateralToWithdrawWei], {
+            gasLimit: 7000000,
+          });
+          setHash(tx.hash as string);
+          await tx.wait();
+        } else if (!resultingCRBelowGCR) {
+          const tx = await emp.withdraw([collateralToWithdrawWei], {
+            gasLimit: 7000000,
+          });
+          setHash(tx.hash as string);
+          await tx.wait();
+        }
+
+        setSuccess(true);
+      } catch (error) {
+        console.error(error);
+        setError(error);
+      }
+    } else {
+      setError(new Error("Please check that you are connected."));
+    }
+  };
+  const executeWithdraw = async () => {
+    if (pendingWithdrawTimeRemaining && emp) {
+      setHash(null);
+      setSuccess(null);
+      setError(null);
+
+      try {
+        const tx = await emp.withdrawPassedRequest({
+          gasLimit: 7000000,
+        });
+        setHash(tx.hash as string);
+        await tx.wait();
+
+        setSuccess(true);
+      } catch (error) {
+        console.error(error);
+        setError(error);
+      }
+    } else {
+      setError(new Error("Please check that you are connected."));
+    }
+  };
+
+  const cancelWithdraw = async () => {
+    if (pendingWithdrawTimeRemaining && emp) {
+      setHash(null);
+      setSuccess(null);
+      setError(null);
+
+      try {
+        const tx = await emp.cancelWithdrawal({
+          gasLimit: 7000000,
+        });
+        setHash(tx.hash as string);
+        await tx.wait();
+
+        setSuccess(true);
+      } catch (error) {
+        console.error(error);
+        setError(error);
+      }
+    } else {
+      setError(new Error("Please check that you are connected."));
+    }
+  };
+
+  const handleWithdrawClick = () => withdrawCollateral();
+  const handleExecuteWithdrawClick = () => executeWithdraw();
+  const handleCancelWithdrawClick = () => cancelWithdraw();
+
+  const startingCR = collateral && tokens ? collateral / tokens : null;
+
+  const resultingCR =
+    collateral && collateralToWithdraw && tokens
+      ? collateral - parseFloat(collateralToWithdraw) / tokens
+      : startingCR;
+
+  const resultingCRBelowGCR = resultingCR && gcr ? resultingCR < gcr : null;
+
+  const fastWithdrawableCollateral =
+    collateral && tokens && gcr ? (collateral - gcr * tokens).toFixed(2) : null;
+
+  const pendingWithdrawTimeRemaining =
```

Comment:
I could not get this to update dynamically. I tried to use a `useEffect` but couldn't get it right.

Reply:
`getPosition` ([here](https://github.com/adrianmcli/emp-tools/blob/6de1114160f9f207ce03f02fca3b4c6ba92533ad/containers/Position.ts#L60)) should be called every time there's a new block. Which means it should cause this component to re-render as well.

Is this not the case?